### PR TITLE
Make State balances Private

### LINKED ISSUE #105: [Driver] Balance list on state struct should be private
This would involve inclusion of public read and write methods balance and removing utility balance_index function.

### PR DESCRIPTION
**Pull request by bh2smith**
closes #105

## Test Plan

Run `cargo test`

### CODE REVIEW

File: driver/src/models.rs [hunk 1]
```diff
@@ -21,9 +21,30 @@ pub trait Serializable {
...
+    pub fn read_balance(&self, token_id: u8, account_id: u16) -> u128 {
+        self.balances[State::balance_index(token_id, account_id)]
+    }
+    pub fn increment_balance(&mut self, token_id: u8, account_id: u16, amount: u128) {
+        self.balances[State::balance_index(token_id, account_id)] += amount;
+    }
+    pub fn decrement_balance(&mut self, token_id: u8, account_id: u16, amount: u128) {
+        self.balances[State::balance_index(token_id, account_id)] -= amount;
+    }
+    pub fn read_balances(&self) -> Vec<u128> {
```
[the remark this answers is not in the document]
Comment (bh2smith) [reply driver/src/models.rs]: I am not sure I understand...

File: driver/src/models.rs [hunk 2]
```diff
@@ -21,9 +21,30 @@ pub trait Serializable {
...
+    pub state_hash: String,
+    pub state_index: i32,
+    balances: Vec<u128>,
+}
+
+impl State {
+    pub fn new(state_hash: String, state_index: i32, balances: Vec<u128>) -> Self {
+        State { state_hash, state_index, balances }
+    }
+    pub fn balance_index(token_id: u8, account_id: u16) -> usize {
```
[the remark this answers is not in the document]
Comment (bh2smith) [reply driver/src/models.rs]: Yes, actually I had intended on making this private, but at some point thought I still needed it. Should be fine to restrict back to private at this point.

File: driver/src/order_driver.rs [hunk 3]
```diff
@@ -102,51 +92,16 @@ mod tests {
     use mock_it::Matcher::*;
     use web3::types::{H256, U256};
 
-    #[test]
-    fn computes_updated_balance_on_example_with_equal_buy_and_sell(){
```
[the remark this answers is not in the document]
Comment (bh2smith) [reply driver/src/order_driver.rs]: Right. I should probably re-introduce a test... I had removed this because the function was removed. Will try and bring it back.

File: driver/src/models.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bh2smith) [reply driver/src/models.rs]:
```suggestion
    fn balance_index(token_id: u8, account_id: u16) -> usize {
```

File: driver/src/models.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (fleupold) [reply driver/src/models.rs]: If I read correctly, the only reason this is needed, is because we want to serialize the balances. We could also expose a `serialize_balances` method directly as part of this trait.

Otherwise, I'd suggest to return a &Vec<u128> here to avoid the clone. 

Or am I missing a use case?

[the remark this answers is not in the document]
Comment (bh2smith) [reply driver/src/models.rs]: Ok I see what you mean now. This is because `read_balances` is only used once (in the same place as `serialize_balances` could be used

File: driver/src/order_driver.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bh2smith) [reply driver/src/order_driver.rs]: Yes! Great, I have reintroduced this test in a more appropriate location (models.py) as a new function implemented for state (update_balances).

File: driver/src/models.rs [hunk 4]
```diff
@@ -21,9 +25,53 @@ pub trait Serializable {
...
+            accounts.insert(account_id(current_account), (0..num_tokens)
+                .map(token_id)
+                .zip(account_balances.iter().map(|b| b.to_string()))
+                .collect());
+            current_account += 1;
+        }
+        json!(accounts)
+    }
+
+    pub fn update_balances(&mut self, orders: &[Order], solution: &Solution) {
```
Comment (fleupold) [inline driver/src/models.rs]: I'm not entirely sure if this method should be part of the State trait. It makes it so that `State` has to know about `Solution`, which is a coupling that might not be necessary. E.g. for deposits and withdraws you apply them inside the driver (that is the `State` trait doesn't have any reference to the `PendingFlux` object).

Anyways, this is just architectural nitty gritty, if it becomes a problem in the future, we can easily put this method back into the driver.

File: driver/src/price_finding/linear_optimization_price_finder.rs [hunk 5]
```diff
@@ -43,11 +43,11 @@ impl LinearOptimisationPriceFinder {
     }
 }
 
-fn token_id(token: u8) -> String {
+pub fn token_id(token: u8) -> String {
```
Comment (fleupold) [inline driver/src/price_finding/linear_optimization_price_finder.rs]: Why is this necessary? I don't see this method being called outside of this model.
